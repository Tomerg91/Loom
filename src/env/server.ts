import { env, type ServerEnv } from './base';

if (typeof window !== 'undefined') {
  throw new Error('serverEnv can only be used in a server-side context');
}

export const serverEnv: ServerEnv = env.server;
