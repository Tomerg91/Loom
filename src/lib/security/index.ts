export * from './headers';
export * from './rate-limit';
export * from './rate-limit-simple';
export * from './validation';
export * from './cors';
export * from './admin-middleware';
export * from './file-security-middleware';

