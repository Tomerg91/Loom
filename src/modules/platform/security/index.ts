/**
 * @fileoverview Public exports for platform security helpers. Keeping exports in a
 * dedicated barrel file ensures future guard utilities remain discoverable while
 * avoiding deep import paths throughout the application.
 */

export * from './httpGuard';
