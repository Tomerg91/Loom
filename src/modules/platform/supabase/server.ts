/**
 * @fileoverview Server-side Supabase factories and guards used by API routes,
 * middleware, and background jobs. The module centralizes cookie-aware client
 * creation and protects service-role access behind explicit helpers so we can
 * safely share the logic across the platform module.
 */
/* eslint-disable @typescript-eslint/no-explicit-any */

import { createServerClient as createSupabaseServerClient } from '@supabase/ssr';
import { createClient as createSupabaseClient } from '@supabase/supabase-js';
import type { NextRequest, NextResponse } from 'next/server';

import {
  PLACEHOLDER_SUPABASE_ANON_KEY,
  PLACEHOLDER_SUPABASE_URL,
} from '@/env/server';

/**
 * Broad Supabase client type. We intentionally keep it loose until the
 * generated database types are refreshed to cover the full schema.
 */
export type ServerSupabaseClient = ReturnType<
  typeof createSupabaseServerClient<any>
>;
export type AdminSupabaseClient = ReturnType<typeof createSupabaseClient<any>>;

type SupabaseCookie = {
  name: string;
  value: string;
  options?: Record<string, unknown> & {
    sameSite?: 'strict' | 'lax' | 'none' | boolean;
  };
};

let serverClientInstance: ServerSupabaseClient | null = null;
let adminClientInstance: AdminSupabaseClient | null = null;

type NextCookieStore = {
  getAll: () => { name: string; value: string }[];
  set?: (
    name: string,
    value: string,
    options?: Record<string, unknown>
  ) => void;
};

/**
 * Validates the required server-side environment variables.
 * Uses process.env directly to work in Edge Runtime (Vercel)
 */
function validateSupabaseEnv(): void {
  // Use process.env directly instead of serverEnv wrapper to avoid
  // ENVIRONMENT_FALLBACK errors in Vercel's Edge Runtime
  const url = process.env.NEXT_PUBLIC_SUPABASE_URL;
  const anonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

  if (
    !url ||
    url === PLACEHOLDER_SUPABASE_URL ||
    url.startsWith('MISSING_') ||
    url.startsWith('INVALID_')
  ) {
    throw new Error(
      'Missing or invalid NEXT_PUBLIC_SUPABASE_URL. Set the project URL before starting the server.'
    );
  }

  try {
    new URL(url);
  } catch (_error) {
    throw new Error(`Invalid Supabase URL format: ${url}`);
  }

  if (!anonKey || anonKey === PLACEHOLDER_SUPABASE_ANON_KEY) {
    throw new Error(
      'Missing or invalid NEXT_PUBLIC_SUPABASE_ANON_KEY. Set the publishable key before starting the server.'
    );
  }
}

/**
 * Creates (or reuses) a server-side Supabase client without cookie context.
 * Useful for scripts or modules that operate outside an HTTP request.
 */
export const createServerClient = (): ServerSupabaseClient => {
  validateSupabaseEnv();

  if (serverClientInstance) {
    return serverClientInstance;
  }

  // Use process.env directly to work in Edge Runtime (Vercel)
  const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!;
  const supabaseKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;


  serverClientInstance = createSupabaseServerClient<any>(
    supabaseUrl,
    supabaseKey,
    {
      cookies: {
        get: () => undefined,
        set: () => {},
        remove: () => {},
      },
    }
  );

  return serverClientInstance;
};

/**
 * Creates a cookie-aware Supabase server client tied to the incoming request
 * and outgoing response. Middleware and API routes should use this so token
 * rotation can set cookies on the fly.
 *
 * Uses process.env directly to work in Vercel's Edge Runtime
 */
export const createServerClientWithRequest = (
  request: NextRequest,
  response: NextResponse
): ServerSupabaseClient => {
  validateSupabaseEnv();

  // Use process.env directly instead of serverEnv wrapper to avoid
  // ENVIRONMENT_FALLBACK errors in Vercel's Edge Runtime
  const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!;
  const supabaseKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;


  return createSupabaseServerClient<any>(supabaseUrl, supabaseKey, {
    cookies: {
      getAll: () => request.cookies.getAll(),
      setAll: (cookies: SupabaseCookie[]) => {
        cookies.forEach(cookie => {
          const { name, value, options } = cookie;
          try {
            const sameSiteValue =
              options?.sameSite === true
                ? ('strict' as const)
                : options?.sameSite === false
                  ? ('none' as const)
                  : (options?.sameSite as
                      | 'strict'
                      | 'lax'
                      | 'none'
                      | undefined);

            response.cookies.set({
              name,
              value,
              ...options,
              sameSite: sameSiteValue,
            });
          } catch (error) {
            console.warn('Failed to set cookie in middleware:', error);
          }
        });
      },
    },
  });
};

/**
 * Creates a Supabase client that leverages Next.js' `cookies()` helper when
 * available. Falls back to a stateless client when not running on the server.
 */
export const createClient = (): ServerSupabaseClient => {
  validateSupabaseEnv();

  const supabaseUrl = serverEnv.NEXT_PUBLIC_SUPABASE_URL!;
  const supabaseKey = serverEnv.NEXT_PUBLIC_SUPABASE_ANON_KEY!;

  let cookieStore: NextCookieStore | null = null;

  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { cookies } = require('next/headers') as {
      cookies: () => NextCookieStore;
    };
    cookieStore = cookies();
  } catch (_error) {
    // Not in a server context; fall back to a stateless adapter.
  }

  const cookieAdapter = cookieStore
    ? {
        getAll: () => {
          try {
            return cookieStore
              .getAll()
              .map(({ name, value }: { name: string; value: string }) => ({
                name,
                value,
              })) as SupabaseCookie[];
          } catch (error) {
            console.warn('Failed to read cookies:', error);
            return [] as SupabaseCookie[];
          }
        },
        setAll: (newCookies: SupabaseCookie[]) => {
          newCookies.forEach(cookie => {
            const { name, value, options } = cookie;
            try {
              const sameSiteValue =
                options?.sameSite === true
                  ? 'strict'
                  : options?.sameSite === false
                    ? 'none'
                    : options?.sameSite;

              cookieStore.set?.(name, value, {
                ...options,
                sameSite: sameSiteValue,
              });
            } catch (error) {
              console.warn('Failed to set cookie:', error);
            }
          });
        },
      }
    : {
        getAll: () => [] as SupabaseCookie[],
        setAll: () => {},
      };

   
  return createSupabaseServerClient<any>(supabaseUrl, supabaseKey, {
    cookies: cookieAdapter,
  });
};

/**
 * Creates a Supabase admin client using the service role key. Access is tightly
 * controlled and the key is never exposed to the browser.
 */
export const createAdminClient = (): AdminSupabaseClient => {
  validateSupabaseEnv();

  if (adminClientInstance) {
    return adminClientInstance;
  }

  const supabaseUrl = serverEnv.NEXT_PUBLIC_SUPABASE_URL!;
  const serviceRoleKey = serverEnv.SUPABASE_SERVICE_ROLE_KEY;

  if (!serviceRoleKey) {
    throw new Error(
      'Missing required environment variable: SUPABASE_SERVICE_ROLE_KEY'
    );
  }

   
  adminClientInstance = createSupabaseClient<any>(supabaseUrl, serviceRoleKey, {
    auth: {
      autoRefreshToken: false,
      persistSession: false,
    },
  });

  return adminClientInstance;
};
