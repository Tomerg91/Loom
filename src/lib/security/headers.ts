import { NextRequest, NextResponse } from 'next/server';

// Security headers configuration
export const SECURITY_HEADERS = {
  // Content Security Policy
  'Content-Security-Policy': [
    "default-src 'self'",
    "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://js.stripe.com https://checkout.stripe.com",
    "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com",
    "font-src 'self' https://fonts.gstatic.com data:",
    "img-src 'self' data: blob: https: http:",
    "media-src 'self' blob:",
    "object-src 'none'",
    "base-uri 'self'",
    "form-action 'self'",
    "frame-ancestors 'none'",
    "connect-src 'self' https://*.supabase.co https://*.supabase.com ws://localhost:* wss://localhost:*",
    "worker-src 'self' blob:",
    "manifest-src 'self'",
  ].join('; '),

  // Security headers
  'X-Frame-Options': 'DENY',
  'X-Content-Type-Options': 'nosniff',
  'X-XSS-Protection': '1; mode=block',
  'Referrer-Policy': 'strict-origin-when-cross-origin',
  'Permissions-Policy': 'camera=(), microphone=(), geolocation=()',
  
  // HSTS (HTTP Strict Transport Security)
  'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload',
  
  // Cross-Origin headers
  'Cross-Origin-Opener-Policy': 'same-origin',
  'Cross-Origin-Embedder-Policy': 'require-corp',
  'Cross-Origin-Resource-Policy': 'same-origin',
};

// Development environment adjustments
const DEV_CSP_OVERRIDES = {
  'Content-Security-Policy': [
    "default-src 'self'",
    "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://js.stripe.com",
    "style-src 'self' 'unsafe-inline'",
    "font-src 'self' data:",
    "img-src 'self' data: blob: https: http:",
    "connect-src 'self' https://*.supabase.co ws://localhost:* wss://localhost:* http://localhost:*",
    "object-src 'none'",
    "base-uri 'self'",
    "form-action 'self'",
    "frame-ancestors 'none'",
  ].join('; '),
};

export function applySecurityHeaders(request: NextRequest, response: NextResponse) {
  const isDevelopment = process.env.NODE_ENV === 'development';
  const headers = isDevelopment 
    ? { ...SECURITY_HEADERS, ...DEV_CSP_OVERRIDES }
    : SECURITY_HEADERS;

  // Apply all security headers
  Object.entries(headers).forEach(([key, value]) => {
    response.headers.set(key, value);
  });

  // Add security headers for API routes
  if (request.nextUrl.pathname.startsWith('/api/')) {
    response.headers.set('Cache-Control', 'no-store, max-age=0');
    response.headers.set('X-Robots-Tag', 'noindex, nofollow');
  }

  return response;
}

// Rate limiting configuration
export const RATE_LIMITS = {
  // Authentication endpoints
  auth: {
    windowMs: 15 * 60 * 1000, // 15 minutes
    max: 5, // 5 attempts per window
    message: 'Too many authentication attempts, please try again later',
  },
  
  // API endpoints
  api: {
    windowMs: 15 * 60 * 1000, // 15 minutes
    max: 100, // 100 requests per window
    message: 'Too many API requests, please try again later',
  },
  
  // Session booking
  booking: {
    windowMs: 5 * 60 * 1000, // 5 minutes
    max: 10, // 10 booking attempts per window
    message: 'Too many booking attempts, please try again later',
  },
};

// Input validation and sanitization
export function sanitizeInput(input: string): string {
  if (typeof input !== 'string') return '';
  
  return input
    .trim()
    .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '') // Remove script tags
    .replace(/javascript:/gi, '') // Remove javascript: URLs
    .replace(/on\w+\s*=/gi, '') // Remove event handlers
    .replace(/data:(?!image\/)/gi, '') // Allow only image data URLs
    .slice(0, 10000); // Limit length
}

// SQL injection prevention patterns
export const SQL_INJECTION_PATTERNS = [
  /(\s|^)(select|insert|update|delete|drop|create|alter|exec|union|script)\s/gi,
  /(\s|^)(or|and)\s+\d+\s*=\s*\d+/gi,
  /(\s|^)(or|and)\s+'.*?'\s*=\s*'.*?'/gi,
  /(\s|^)(or|and)\s+".*?"\s*=\s*".*?"/gi,
  /--/g,
  /\/\*/g,
  /\*\//g,
];

export function containsSQLInjection(input: string): boolean {
  return SQL_INJECTION_PATTERNS.some(pattern => pattern.test(input));
}

// XSS prevention patterns
export const XSS_PATTERNS = [
  /<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi,
  /javascript:/gi,
  /on\w+\s*=/gi,
  /<iframe\b[^<]*(?:(?!<\/iframe>)<[^<]*)*<\/iframe>/gi,
  /<object\b[^<]*(?:(?!<\/object>)<[^<]*)*<\/object>/gi,
  /<embed\b[^<]*(?:(?!<\/embed>)<[^<]*)*<\/embed>/gi,
  /<svg\b[^<]*(?:(?!<\/svg>)<[^<]*)*<\/svg>/gi,
];

export function containsXSS(input: string): boolean {
  return XSS_PATTERNS.some(pattern => pattern.test(input));
}

// XSS prevention
export function sanitizeHTML(html: string): string {
  // Simple HTML sanitization - in production, use a library like DOMPurify
  return html
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;')
    .replace(/\//g, '&#x2F;');
}

// CSRF token generation and validation
export function generateCSRFToken(): string {
  return Array.from(crypto.getRandomValues(new Uint8Array(32)))
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');
}

export function validateCSRFToken(token: string, sessionToken: string): boolean {
  return token === sessionToken && token.length === 64;
}

// Password strength validation
export function validatePasswordStrength(password: string): {
  isValid: boolean;
  errors: string[];
} {
  const errors: string[] = [];
  
  if (password.length < 8) {
    errors.push('Password must be at least 8 characters long');
  }
  
  if (!/[A-Z]/.test(password)) {
    errors.push('Password must contain at least one uppercase letter');
  }
  
  if (!/[a-z]/.test(password)) {
    errors.push('Password must contain at least one lowercase letter');
  }
  
  if (!/\d/.test(password)) {
    errors.push('Password must contain at least one number');
  }
  
  if (!/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
    errors.push('Password must contain at least one special character');
  }
  
  // Check for common weak patterns
  const weakPatterns = [
    /(.)\1{2,}/, // Repeated characters
    /123456/, // Sequential numbers
    /qwerty/i, // Common keyboard patterns
    /password/i, // Common words
    /admin/i,
    /user/i,
  ];
  
  if (weakPatterns.some(pattern => pattern.test(password))) {
    errors.push('Password contains common weak patterns');
  }
  
  return {
    isValid: errors.length === 0,
    errors,
  };
}

// File upload security
export const ALLOWED_FILE_TYPES = {
  images: ['image/jpeg', 'image/png', 'image/gif', 'image/webp'],
  documents: ['application/pdf', 'text/plain'],
};

export const MAX_FILE_SIZES = {
  avatar: 5 * 1024 * 1024, // 5MB
  document: 10 * 1024 * 1024, // 10MB
};

export function validateFileUpload(file: File, type: 'avatar' | 'document'): {
  isValid: boolean;
  errors: string[];
} {
  const errors: string[] = [];
  
  // Check file type
  const allowedTypes = type === 'avatar' ? ALLOWED_FILE_TYPES.images : ALLOWED_FILE_TYPES.documents;
  if (!allowedTypes.includes(file.type)) {
    errors.push(`Invalid file type. Allowed types: ${allowedTypes.join(', ')}`);
  }
  
  // Check file size
  const maxSize = MAX_FILE_SIZES[type];
  if (file.size > maxSize) {
    errors.push(`File too large. Maximum size: ${Math.round(maxSize / (1024 * 1024))}MB`);
  }
  
  // Check filename for suspicious patterns
  const suspiciousPatterns = [
    /\.(exe|bat|cmd|scr|pif|com|dll)$/i,
    /\.(php|jsp|asp|js|html)$/i,
    /\.\./, // Path traversal
    /%/, // URL encoding
  ];
  
  if (suspiciousPatterns.some(pattern => pattern.test(file.name))) {
    errors.push('Suspicious filename detected');
  }
  
  return {
    isValid: errors.length === 0,
    errors,
  };
}