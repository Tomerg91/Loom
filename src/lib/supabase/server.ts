import { createServerClient as createSupabaseServerClient } from '@supabase/ssr';
import { createClient as createSupabaseClient } from '@supabase/supabase-js';
import { type NextRequest, type NextResponse } from 'next/server';
import { type Database } from '@/types/supabase';
import { env } from '@/env.mjs';
import { serverEnv } from '@/env-server.mjs';

// Singleton instances to prevent multiple GoTrueClient creation
let serverClientInstance: ReturnType<typeof createSupabaseServerClient<Database>> | null = null;
let adminClientInstance: ReturnType<typeof createSupabaseClient<Database>> | null = null;

// Validate required environment variables
function validateSupabaseEnv() {
  if (!env.NEXT_PUBLIC_SUPABASE_URL) {
    throw new Error('Missing required environment variable: NEXT_PUBLIC_SUPABASE_URL');
  }
  if (!env.NEXT_PUBLIC_SUPABASE_ANON_KEY) {
    throw new Error('Missing required environment variable: NEXT_PUBLIC_SUPABASE_ANON_KEY');
  }
  
  // Validate URL format
  if (env.NEXT_PUBLIC_SUPABASE_URL.startsWith('MISSING_') || env.NEXT_PUBLIC_SUPABASE_URL.startsWith('INVALID_')) {
    throw new Error(`Invalid Supabase URL configuration: ${env.NEXT_PUBLIC_SUPABASE_URL}`);
  }
  
  // Validate URL is actually a valid URL
  try {
    new URL(env.NEXT_PUBLIC_SUPABASE_URL);
  } catch (error) {
    throw new Error(`Invalid Supabase URL format: ${env.NEXT_PUBLIC_SUPABASE_URL}`);
  }
}

// Server-side Supabase client for middleware (without cookie access)
export const createServerClient = () => {
  validateSupabaseEnv();
  
  if (serverClientInstance) {
    return serverClientInstance;
  }
  
  const supabaseUrl = env.NEXT_PUBLIC_SUPABASE_URL!;
  const supabaseKey = env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;
  
  serverClientInstance = createSupabaseServerClient<Database>(
    supabaseUrl,
    supabaseKey,
    {
      cookies: {
        get: () => undefined,
        set: () => {},
        remove: () => {},
      },
    }
  );
  
  return serverClientInstance;
};

// Server-side Supabase client for middleware with request context
export const createServerClientWithRequest = (request: NextRequest, response: NextResponse) => {
  validateSupabaseEnv();
  
  const supabaseUrl = env.NEXT_PUBLIC_SUPABASE_URL!;
  const supabaseKey = env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;
  
  return createSupabaseServerClient<Database>(
    supabaseUrl,
    supabaseKey,
    {
      cookies: {
        getAll: () => request.cookies.getAll(),
        setAll: (cookies) => {
          cookies.forEach(({ name, value, options }) => {
            try {
              const sameSiteValue = options?.sameSite === true 
                ? 'strict' as const
                : (options?.sameSite === false 
                    ? 'none' as const 
                    : options?.sameSite as 'strict' | 'lax' | 'none' | undefined);
              
              response.cookies.set({
                name,
                value,
                ...options,
                sameSite: sameSiteValue
              });
            } catch (error) {
              console.warn('Failed to set cookie in middleware:', error);
            }
          });
        },
      },
    }
  );
};

// For route handlers and server components that have access to cookies
export const createClient = async () => {
  validateSupabaseEnv();
  
  const supabaseUrl = env.NEXT_PUBLIC_SUPABASE_URL!;
  const supabaseKey = env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;
  
  // Check if we're in a server context with cookies available
  try {
    // Dynamic import to avoid issues during build
    const { cookies } = await import('next/headers');
    const cookieStore = await cookies();
    
    return createSupabaseServerClient<Database>(
      supabaseUrl,
      supabaseKey,
      {
        cookies: {
          getAll: () => {
            try {
              return cookieStore.getAll();
            } catch (error) {
              console.warn('Failed to get cookies:', error);
              return [];
            }
          },
          setAll: (cookies) => {
            cookies.forEach(({ name, value, options }) => {
              try {
                const sameSiteValue = options?.sameSite === true 
                  ? 'strict' as const
                  : (options?.sameSite === false 
                      ? 'none' as const 
                      : options?.sameSite as 'strict' | 'lax' | 'none' | undefined);
                
                cookieStore.set(name, value, {
                  ...options,
                  sameSite: sameSiteValue
                });
              } catch (error) {
                console.warn('Failed to set cookie:', error);
              }
            });
          },
        },
      }
    );
  } catch (error) {
    // Fallback for build-time or non-server contexts
    console.warn('Cookies not available, falling back to cookieless client:', error);
    return createSupabaseServerClient<Database>(
      supabaseUrl,
      supabaseKey,
      {
        cookies: {
          getAll: () => [],
          setAll: () => {},
        },
      }
    );
  }
};

// Admin client with service role key for administrative operations
export const createAdminClient = () => {
  validateSupabaseEnv();
  
  if (adminClientInstance) {
    return adminClientInstance;
  }
  
  const supabaseUrl = env.NEXT_PUBLIC_SUPABASE_URL!;
  const serviceRoleKey = serverEnv.SUPABASE_SERVICE_ROLE_KEY;
  
  if (!serviceRoleKey) {
    throw new Error('Missing required environment variable: SUPABASE_SERVICE_ROLE_KEY');
  }
  
  adminClientInstance = createSupabaseClient<Database>(supabaseUrl, serviceRoleKey, {
    auth: {
      autoRefreshToken: false,
      persistSession: false,
    },
  });
  
  return adminClientInstance;
};