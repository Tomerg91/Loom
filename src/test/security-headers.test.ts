import { NextRequest, NextResponse } from 'next/server';
import { describe, it, expect } from 'vitest';

import { applySecurityHeaders } from '@/lib/security/headers';

describe('Security Headers - Snapshot Tests', () => {
  it('should apply correct security headers to HTML responses', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();
    res.headers.set('content-type', 'text/html');

    const secureRes = applySecurityHeaders(req, res);

    const headers: Record<string, string> = {};
    secureRes.headers.forEach((value, key) => {
      // Normalize nonce values for snapshot testing
      if (key.toLowerCase() === 'content-security-policy') {
        headers[key] = value.replace(/nonce-[a-zA-Z0-9+/=]+/g, 'nonce-NORMALIZED');
      } else {
        headers[key] = value;
      }
    });

    expect(headers).toMatchSnapshot();
  });

  it('should apply correct security headers to JSON API responses', () => {
    const req = new NextRequest(new URL('http://localhost:3000/api/sessions'));
    const res = NextResponse.json({ data: [] });

    const secureRes = applySecurityHeaders(req, res);

    const headers: Record<string, string> = {};
    secureRes.headers.forEach((value, key) => {
      // Normalize nonce values for snapshot testing
      if (key.toLowerCase() === 'content-security-policy') {
        headers[key] = value.replace(/nonce-[a-zA-Z0-9+/=]+/g, 'nonce-NORMALIZED');
      } else {
        headers[key] = value;
      }
    });

    expect(headers).toMatchSnapshot();
  });

  it('should include X-Frame-Options header', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);

    expect(secureRes.headers.has('X-Frame-Options')).toBe(true);
    expect(secureRes.headers.get('X-Frame-Options')).toBe('DENY');
  });

  it('should include X-Content-Type-Options header', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);

    expect(secureRes.headers.has('X-Content-Type-Options')).toBe(true);
    expect(secureRes.headers.get('X-Content-Type-Options')).toBe('nosniff');
  });

  it('should include Referrer-Policy header', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);

    expect(secureRes.headers.has('Referrer-Policy')).toBe(true);
    expect(secureRes.headers.get('Referrer-Policy')).toBe('strict-origin-when-cross-origin');
  });

  it('should include Permissions-Policy header', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);

    expect(secureRes.headers.has('Permissions-Policy')).toBe(true);
  });

  it('should include Strict-Transport-Security header in production', () => {
    const originalEnv = process.env.NODE_ENV;
    vi.stubEnv('NODE_ENV', 'production');

    const req = new NextRequest(new URL('https://example.com/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);

    expect(secureRes.headers.has('Strict-Transport-Security')).toBe(true);
    expect(secureRes.headers.get('Strict-Transport-Security')).toContain('max-age=');

    process.env.NODE_ENV = originalEnv;
  });

  it('should include Content-Security-Policy header', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);

    expect(secureRes.headers.has('Content-Security-Policy')).toBe(true);
    const csp = secureRes.headers.get('Content-Security-Policy');
    expect(csp).toContain("default-src 'self'");
  });

  it('should generate unique nonce for each request when CSP_NONCE_ENABLED=true', () => {
    const originalEnv = process.env.CSP_NONCE_ENABLED;
    const originalNodeEnv = process.env.NODE_ENV;
    process.env.CSP_NONCE_ENABLED = 'true';
    vi.stubEnv('NODE_ENV', 'production');

    const req1 = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res1 = NextResponse.next();
    const secureRes1 = applySecurityHeaders(req1, res1);

    const req2 = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res2 = NextResponse.next();
    const secureRes2 = applySecurityHeaders(req2, res2);

    const csp1 = secureRes1.headers.get('Content-Security-Policy') || '';
    const csp2 = secureRes2.headers.get('Content-Security-Policy') || '';

    const nonce1Match = csp1.match(/nonce-([a-zA-Z0-9+/=]+)/);
    const nonce2Match = csp2.match(/nonce-([a-zA-Z0-9+/=]+)/);

    expect(nonce1Match).toBeTruthy();
    expect(nonce2Match).toBeTruthy();
    expect(nonce1Match?.[1]).not.toBe(nonce2Match?.[1]);

    process.env.CSP_NONCE_ENABLED = originalEnv;
    process.env.NODE_ENV = originalNodeEnv;
  });
});

describe('Security Headers - CSP Directives', () => {
  it('should allow Supabase domains in CSP', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);
    const csp = secureRes.headers.get('Content-Security-Policy') || '';

    expect(csp).toContain('*.supabase.co');
  });

  it('should allow Vercel Live in CSP', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);
    const csp = secureRes.headers.get('Content-Security-Policy') || '';

    expect(csp).toContain('vercel.live');
  });

  it('should have strict default-src policy', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);
    const csp = secureRes.headers.get('Content-Security-Policy') || '';

    expect(csp).toMatch(/default-src[^;]*'self'/);
  });

  it('should allow inline scripts with nonce when CSP_NONCE_ENABLED=true', () => {
    const originalEnv = process.env.CSP_NONCE_ENABLED;
    const originalNodeEnv = process.env.NODE_ENV;
    process.env.CSP_NONCE_ENABLED = 'true';
    vi.stubEnv('NODE_ENV', 'production');

    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);
    const csp = secureRes.headers.get('Content-Security-Policy') || '';

    expect(csp).toMatch(/script-src[^;]*'nonce-/);

    process.env.CSP_NONCE_ENABLED = originalEnv;
    process.env.NODE_ENV = originalNodeEnv;
  });

  it('should allow unsafe-inline scripts by default (when nonce disabled)', () => {
    const req = new NextRequest(new URL('http://localhost:3000/en/dashboard'));
    const res = NextResponse.next();

    const secureRes = applySecurityHeaders(req, res);
    const csp = secureRes.headers.get('Content-Security-Policy') || '';

    expect(csp).toMatch(/script-src[^;]*'unsafe-inline'/);
  });
});
