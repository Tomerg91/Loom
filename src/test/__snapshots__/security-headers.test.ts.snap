// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Security Headers - Snapshot Tests > should apply correct security headers to HTML responses 1`] = `
{
  "content-security-policy": "script-src 'self' 'unsafe-inline' https://vercel.live https://secure5.tranzila.com https://direct.tranzila.com https://js.sentry-cdn.com https://*.sentry.io; default-src 'self'; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com data:; img-src 'self' data: blob: https:; media-src 'self' blob:; object-src 'none'; base-uri 'self'; form-action 'self' https://secure5.tranzila.com https://direct.tranzila.com; frame-ancestors 'none'; frame-src 'self' https://vercel.live https://secure5.tranzila.com https://direct.tranzila.com https://*.sentry.io; connect-src 'self' https://*.supabase.co https://*.supabase.com wss://*.supabase.co wss://*.supabase.com https://vercel.live wss://vercel.live https://secure5.tranzila.com https://direct.tranzila.com https://sentry.io https://*.sentry.io; worker-src 'self' blob:; manifest-src 'self'",
  "content-type": "text/html",
  "cross-origin-embedder-policy": "unsafe-none",
  "cross-origin-opener-policy": "same-origin-allow-popups",
  "cross-origin-resource-policy": "cross-origin",
  "permissions-policy": "camera=(), microphone=(), geolocation=()",
  "referrer-policy": "strict-origin-when-cross-origin",
  "strict-transport-security": "max-age=31536000; includeSubDomains; preload",
  "x-content-type-options": "nosniff",
  "x-frame-options": "DENY",
  "x-middleware-next": "1",
}
`;

exports[`Security Headers - Snapshot Tests > should apply correct security headers to JSON API responses 1`] = `
{
  "cache-control": "no-store, max-age=0",
  "content-security-policy": "script-src 'self' 'unsafe-inline' https://vercel.live https://secure5.tranzila.com https://direct.tranzila.com https://js.sentry-cdn.com https://*.sentry.io; default-src 'self'; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com data:; img-src 'self' data: blob: https:; media-src 'self' blob:; object-src 'none'; base-uri 'self'; form-action 'self' https://secure5.tranzila.com https://direct.tranzila.com; frame-ancestors 'none'; frame-src 'self' https://vercel.live https://secure5.tranzila.com https://direct.tranzila.com https://*.sentry.io; connect-src 'self' https://*.supabase.co https://*.supabase.com wss://*.supabase.co wss://*.supabase.com https://vercel.live wss://vercel.live https://secure5.tranzila.com https://direct.tranzila.com https://sentry.io https://*.sentry.io; worker-src 'self' blob:; manifest-src 'self'",
  "content-type": "application/json",
  "cross-origin-embedder-policy": "unsafe-none",
  "cross-origin-opener-policy": "same-origin-allow-popups",
  "cross-origin-resource-policy": "cross-origin",
  "permissions-policy": "camera=(), microphone=(), geolocation=()",
  "referrer-policy": "strict-origin-when-cross-origin",
  "strict-transport-security": "max-age=31536000; includeSubDomains; preload",
  "x-content-type-options": "nosniff",
  "x-frame-options": "DENY",
  "x-robots-tag": "noindex, nofollow",
}
`;
